### PR TITLE
fix: terminal/task UI and many many terminal functionality changes

### LINKED ISSUE #109: Terminal Functionality
Todo:
- [x] mv
- [x] grep
- [x] find
- [x] chmod
- [x] cat with piping/redirection
- [x] longFormat for ``ls -l`` in globalTypes.ts
- [x] ``cp file1 file2 directoryPath``

### PR DESCRIPTION
**Pull request by jason2020**
## Summary

Closes #109 

### Terminal/Task UI changes:
- Replaced all terminals across all pages with Task component 
   - Manually copied over styling from #110  (Lmk if I missed anything!)
- Removed unnecessary CSS (the preexisting Task component had most of the styling)
- Added option to hide the file system render for pages like Stationary

### Terminal functionality changes:
- Added rm, rmdir, cp (also includes rm -rf and cp -R)
- Fixed issues with relative paths not working
- Added clear
- Added better flag handling (e.g. ls -a -l now works), this will make future changes easier
- Added feature to cycle through the command history (using the up and down arrow keys)

### TODOs (can be future tickets):
- Finish mv, grep, find, chmod
   - Add in permissions to the file system
- Add piping and redirection
- ls long format (requires adding additional metadata to file system objects)
- Fix issue with cp overwriting dir
- Clean up code (remove repeated code using helper functions like getFSObjectHelper)
- Fix styling for ice-glare-left if displaying the file system render
- Add in the fileSystem instances for all the terminals across the different pages

### Stationary
Screen Shot 2022-09-28 at 1 23 55 PM
Screen Shot 2022-09-28 at 1 24 27 PM

### Moving
Screen Shot 2022-09-28 at 1 25 21 PM

### Piping
Screen Shot 2022-09-28 at 1 25 53 PM
Screen Shot 2022-09-28 at 1 26 05 PM

### Recording of some of the terminal functionality
https://user-images.githubusercontent.com/13056466/192882928-8226357f-4692-4f53-b1d4-ac5c8a41d227.mov

https://user-images.githubusercontent.com/13056466/192883428-9b0f8cdf-116d-4ca9-af5f-8a9e89369813.mov

### CODE REVIEW

File: src/components/shared/Terminal.tsx [hunk 1]
```diff
@@ -12,11 +12,13 @@ function Terminal(prop: {
...
   const [commands, setCommands] = useState<string[]>([]);
   const [input, setInput] = useState('$ ');
+  const [inputHistory, setInputHistory] = useState<string[]>([]);
+  const [inputHistoryIndex, setInputHistoryIndex] = useState(0);
 
   function executeList(flags: string, path: string): [string[], string[]] {
     const displayAll = flags.includes('a');
     const longFormat = flags.includes('l');
-    const fsObject = path.includes('/')
+    const fsObject = path.startsWith('/')
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:21]: Refactor - make these strings for path and file into constants for reusability

File: src/components/shared/Terminal.tsx [hunk 2]
```diff
@@ -64,18 +66,25 @@ function Terminal(prop: {
   }
 
   function executeTouch(path: string, directory = false): string[] {
+    // Make a copy to avoid mutating the original (it's a state variable)
     const fileSystemCopy = _.cloneDeep(prop.fileSystem);
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:70]: Refactor - could we rename this to rootCopy so the naming is less confusing? Thanks 🙂!

File: src/components/shared/Terminal.tsx [hunk 3]
```diff
@@ -64,18 +66,25 @@ function Terminal(prop: {
...
     const cwdCopy = fileSystemCopy.getFileSystemObjectFromPath(
       prop.currentWorkingDirectory.path
     );
     let newFileName = '';
     let dir;
 
+    // If the path is absolute, find the directory to create the file in
+    // Otherwise, use the current working directory
     if (path.includes('/')) {
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:79]: Refactor/backlog - Could we make this into a helper function to make it more readable / reusable

File: src/components/shared/Terminal.tsx [hunk 4]
```diff
@@ -104,15 +113,131 @@ function Terminal(prop: {
     return executeTouch(path, true);
   }
 
+  function executeRm(path: string, flags: string, directory = false): string[] {
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:116]: Add a comment above this function explaining what it does -- is it the same as `rm`? Does it remove directories? Etc. Just a high-level comment will suffice 🙂

File: src/components/shared/Terminal.tsx [hunk 5]
```diff
@@ -104,15 +113,131 @@ function Terminal(prop: {
...
+    } else {
+      if (directory && !file.isEmpty()) {
+        return [`rmdir: '${path}': Directory not empty`];
+      }
+      return [`rm: '${path}': is a directory`];
+    }
+
+    prop.setFileSystem(fileSystemCopy);
+    prop.setCurrentWorkingDirectory(cwdCopy);
+    return [];
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:164]: Refactor - replace this with a constant that's same as empty array for no errors. For example, what if we want to have a debug mode that does show "returned with success"?

File: src/components/shared/Terminal.tsx [hunk 6]
```diff
@@ -104,15 +113,131 @@ function Terminal(prop: {
...
+      dir = cwdCopy;
+    }
+
+    const file = dir?.getChild(fileName);
+    if (!file) {
+      return [`rm: '${path}': No such file or directory`];
+    }
+
+    if (
+      (directory && file.isEmpty()) ||
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:146]: Could we make each of these cases into a constant with the name explaining what it does? Kind of confused here lol

File: src/components/shared/Terminal.tsx [hunk 7]
```diff
@@ -104,15 +113,131 @@ function Terminal(prop: {
...
+  }
+
+  function executeRmdir(path: string, flags: string): string[] {
+    return executeRm(path, flags, true);
+  }
+
+  function executeCopy(
+    path: string,
+    destination: string,
+    flags: string
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:174]: Refactor - make the flags into an array of strings instead of string

File: src/components/shared/Terminal.tsx [hunk 8]
```diff
@@ -104,15 +113,131 @@ function Terminal(prop: {
...
+        destDir = fileSystemCopy.getFileSystemObjectFromPath(pathArr.join('/'));
+      } else {
+        destDir = cwdCopy.getFileSystemObjectFromPath(pathArr.join('/'));
+      }
+    } else {
+      newFileName = destination;
+      destDir = cwdCopy;
+    }
+
+    if (!file) {
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:202]: Refactor - upgrade this check to be `file === null`. It'd be better if we explicitly check the lack of file instead of just handling undefined

File: src/components/shared/Terminal.tsx [hunk 9]
```diff
@@ -104,15 +113,131 @@ function Terminal(prop: {
...
+      return [`cp: '${path}': No such file or directory`];
+    } else if (file.isDirectory && !flags.toLowerCase().includes('r')) {
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:204]: Actually, you don't even need the `else if` here, you can just do `if` since it's a guard clause

File: src/components/shared/Terminal.tsx [hunk 10]
```diff
@@ -104,15 +113,131 @@ function Terminal(prop: {
...
+    newFile.name = newFileName;
+    destDir.addFileSystemObject(newFile);
+    prop.setFileSystem(fileSystemCopy);
+    prop.setCurrentWorkingDirectory(cwdCopy);
+    return [];
+  }
+
+  function executeFind(path: string, flags: string): [string[], string[]] {
+    if (path || flags) {
+      return [[], ['find: Not implemented']];
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:221]: Refactor - make this into an object with stderr and stdout

File: src/components/shared/Terminal.tsx [hunk 11]
```diff
@@ -104,15 +113,131 @@ function Terminal(prop: {
...
   function isValid(usertyped: string) {
     const userInput = usertyped.trim().split(' ');
     const command = input.length > 1 ? userInput.slice(1, 2).join(' ') : '';
     const args = userInput.slice(2);
-    const flags = args.length > 0 && args[0].startsWith('-') ? args[0] : '';
-    const path =
-      args.length > 0 && args[0].startsWith('-')
-        ? args.slice(1).join(' ')
-        : args.join(' ');
+    let flags = '';
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:230]: 1. TO-DO - Flags should be an array of strings instead of one string
2. TO-DO - Set up parsing for the flags - please make this in a separate PR with tests (please!)
3. Consideration - might want to refactor out the loop using indexOf() even though performance is very very slightly slower but it'll be a lot more readable (also your loop has mutations and saving state (i))

File: src/components/shared/Terminal.tsx [hunk 12]
```diff
@@ -130,29 +255,32 @@ function Terminal(prop: {
         console.log('User input man whoami');
         break;
       case 'ls':
-        [output, error] = executeList(flags, path === '' ? '.' : path);
+        [output, error] = executeList(flags, path);
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:258]: Nice job on cleanup! 😄

File: src/components/shared/globalTypes.ts [hunk 13]
```diff
@@ -153,31 +156,44 @@ export class Directory extends FileSystemObject {
   changeCurrentWorkingDirectory(
     currentWorkingDirectory: Directory,
     path: string
-  ): Directory | undefined {
```
Comment (RogueArt) [inline src/components/shared/globalTypes.ts:156]: What????

File: src/components/shared/globalTypes.ts [hunk 14]
```diff
@@ -153,31 +156,44 @@ export class Directory extends FileSystemObject {
...
+    // If the path is absolute, we start from the root directory and find the new cwd
+    if (path.startsWith('/')) {
       newCwd = this.getFileSystemObjectFromPath(path);
+    } else {
+      newCwd = currentWorkingDirectory.getFileSystemObjectFromPath(path);
     }
 
-    if (!newCwd || !newCwd.isDirectory) return undefined;
+    // If the new cwd is a file or does not exist, we can't cd into it
+    if (!newCwd) {
```
Comment (RogueArt) [inline src/components/shared/globalTypes.ts:171]: Couldn't these two `if` statements be combined into one?

File: src/components/shared/globalTypes.ts [hunk 15]
```diff
@@ -153,31 +156,44 @@ export class Directory extends FileSystemObject {
...
+    }
     (newCwd as Directory).isCurrentDirectory = true;
     return newCwd as Directory;
   }
 
   getFileSystemObjectFromPath(path: string): Directory | File | undefined {
     if (path === this.path) return this;
 
+    // Removes the trailing slash if it exists
     if (path.charAt(path.length - 1) == '/') {
```
Comment (RogueArt) [inline src/components/shared/globalTypes.ts:185]: Replace with `path.endsWith('/')`?

File: src/components/shared/globalTypes.ts [hunk 16]
```diff
@@ -153,31 +156,44 @@ export class Directory extends FileSystemObject {
...
       path = path.slice(0, -1);
     }
 
-    const children = path.includes('/') ? path.split('/').slice(1) : [path];
+    const children = path.includes('/')
+      ? path.split('/').slice(path.startsWith('/') ? 1 : 0)
```
Comment (RogueArt) [inline src/components/shared/globalTypes.ts:190]: Could do `if (path.startsWith'/') path.replace('/', '')`

File: src/components/shared/globalTypes.ts [hunk 17]
```diff
@@ -153,31 +156,44 @@ export class Directory extends FileSystemObject {
...
+      : [path];
+
     let currentDirectory = <Directory>this;
     let currentFsObject: Directory | File | undefined;
+
+    // Traverse the path to find the new cwd and return the fs object
     for (let i = 0; i < children.length; i++) {
```
Comment (RogueArt) [inline src/components/shared/globalTypes.ts:197]: Replace it with this for...of loop - `for (const child of children)`

File: src/components/shared/globalTypes.ts [hunk 18]
```diff
@@ -195,7 +211,31 @@ export class Directory extends FileSystemObject {
     return currentFsObject;
   }
 
+  isEmpty(): boolean {
+    return this.children === undefined || this.children.size === 0;
```
Comment (RogueArt) [inline src/components/shared/globalTypes.ts:215]: Which language are you using? `.size`?

File: src/components/shared/globalTypes.ts [hunk 19]
```diff
@@ -195,7 +211,31 @@ export class Directory extends FileSystemObject {
...
     return name.startsWith('.');
   }
 }
+
+export function getFSObjectHelper(
+  path: string,
+  root: Directory,
+  cwd: Directory,
+  onFileNotFound?: () => string,
+  onIsNotDirectory?: () => string
```
Comment (RogueArt) [inline src/components/shared/globalTypes.ts:228]: Break these up into separate helpers
- `fileNotFoundHelper(onFileNotFound: () => string)`
- `fileIsNotDirectoryHelper(onIsNotDirectory: () => string)`
And then call the master helper function on the inside. This way we enforce the `null` check without having to explicitly check for it

File: src/styles/Task.scss [hunk 20]
```diff
@@ -2,10 +2,11 @@
   border-radius: 10px;
   display: flex;
   flex-direction: column;
-  margin: 1vh auto;
-  margin-bottom: 5vh;
+  margin: 0 132px;
```
Comment (RogueArt) [inline src/styles/Task.scss:5]: 🤔

File: src/components/shared/Terminal.tsx [hunk 21]
```diff
@@ -130,29 +255,32 @@ function Terminal(prop: {
...
-        console.log('User input rmdir');
+        error = executeRmdir(path, flags);
         break;
-      case 'rm -rf':
-        console.log('User input rm -rf');
-        break;
-      case 'cp':
-        console.log('User input cp');
+      case 'cp': {
+        const cpArgs = path.split(' ');
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:276]: 1. TO-DO - Please replace with the regex `/\s+/` to handle arbitrary whitespace
2. TO-DO - Move this check to earlier so we handle it once and done

File: src/components/shared/Terminal.tsx [hunk 22]
```diff
@@ -191,10 +322,39 @@ function Terminal(prop: {
   const handleSubmit = (e: React.FormEvent) => {
     e.preventDefault();
     const result = isValid(input);
-    setCommands([...commands, input, ...result.err, ...result.out]);
+    const newHistory = [...inputHistory, input];
+    if (result.err.length > 0 && result.err[0] === 'CLEAR') {
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:326]: 'CLEAR' should be a constant to avoid people from making typos regarding case

File: src/components/shared/Terminal.tsx [hunk 23]
```diff
@@ -191,10 +322,39 @@ function Terminal(prop: {
...
+      setCommands([]);
+    } else {
+      setCommands([...commands, input, ...result.err, ...result.out]);
+    }
+    setInputHistory(newHistory);
+    setInputHistoryIndex(newHistory.length);
     setInput('$ ');
   };
 
+  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
```
Comment (RogueArt) [inline src/components/shared/Terminal.tsx:336]: For code style, we should make this into a normal function / shouldn't be an arrow (anonymous) function

File: src/components/shared/globalTypes.ts (hunk 19 above)
Comment (RogueArt) [reply src/components/shared/globalTypes.ts:228]: Here
```js
interface ErrorObj {
  onFileNotFound?: () => string;
  onIsNotDirectory?: () => string;
}

export function getFSObjectMasterHelper(
  path: string,
  root: Directory,
  cwd: Directory,
  errorObj?: ErrorObj,
): Directory | File | string {
  const fsObject = path.startsWith('/')
    ? root.getFileSystemObjectFromPath(path)
    : cwd.getFileSystemObjectFromPath(path);

  // Return fsObject in the case of nullish or empty errorObj
  if (!errorObj || Object.keys(errorObj).length === 0) {
    return fsObject;
  }

  if (!fsObject && errorObj.onFileNotFound) {
    return errorObj.onFileNotFound()
  }

  if (errorObj.onIsNotDirectory && !fsObject?.isDirectory) {
    return errorObj.onIsNotDirectory();
  }

  // So technically, it's impossible to hit this case
  // but if they do, let's just throw an exception just in case
  throw new Error("I am not sure what you're doing here tbh, please double check?")
}
```

**Review by RogueArt — CHANGES_REQUESTED** [review]

File: package.json [hunk 24]
```diff
@@ -31,7 +31,6 @@
     "@babel/preset-env": "^7.16.7",
     "@babel/preset-react": "^7.16.7",
     "@babel/preset-typescript": "^7.14.5",
-    "@types/lodash": "^4.14.184",
```
[the remark this answers is not in the document]
Comment (jason2020) [reply package.json:34]: Good catch! I think I missed the change when merging.

File: src/components/pages/moving.tsx [hunk 25]
```diff
@@ -46,281 +47,31 @@ function Moving(): JSX.Element {
...
+          <p className="body task-prompt">
+            Something seems amiss in this cave. Is there a secret hiding in
+            plain sight?
+          </p>
+        }
+        taskName="Task 2"
+        completed={false}
+        fileSystem={undefined}
+        currentWorkingDirectory={undefined}
+      ></Task>
```
[the remark this answers is not in the document]
Comment (jason2020) [reply src/components/pages/moving.tsx:74]: Agreed, I think I'll likely close this PR and refactor code into 1 PR for Task refactor and another for the Terminal functionality. Then, I'll make PR(s) for the backlog tasks resolving some of the issues you mentioned.

[the remark this answers is not in the document]
Comment (jason2020) [reply src/components/pages/moving.tsx:74]: Also, probably a future PR is urgently needed for unit tests as the current process of manual checking for terminal functionality is not very effective or sustainable as this project scales.

File: src/components/shared/Task.tsx [hunk 26]
```diff
@@ -13,6 +13,7 @@ function Task(prop: {
   completed: boolean;
   fileSystem?: Directory;
   currentWorkingDirectory?: Directory;
+  displayFileSystem?: boolean;
```
[the remark this answers is not in the document]
Comment (jason2020) [reply src/components/shared/Task.tsx:16]: Yup makes sense, this will be in PR after terminal func changes and the task fix.

File: src/components/shared/Task.tsx [hunk 27]
```diff
@@ -77,6 +78,90 @@ function Task(prop: {
           setFileSystem={setFileSystem}
           setCurrentWorkingDirectory={setCurrentWorkingDirectory}
         />
+        {/* TODO: Fix styling for ice-glare-left if displaying the file system render */}
+        <div className="ice-glare-left">
```
[the remark this answers is not in the document]
Comment (jason2020) [reply src/components/shared/Task.tsx:82]: Good point, yeah this should improve the code readability and also makes sense if a larger glare left is needed for when the file system render is shown

File: src/components/shared/globalTypes.ts (hunk 19 above)
Comment (jason2020) [reply src/components/shared/globalTypes.ts:228]: I see, yeah this makes sense. I like the idea of using a new interface to handle errors. I will include this in a constants file. Also, will make the change to getFileSystemObjectFromPath to return null rather than undefined, which should work with `!fsObject`.

Comment (jason2020) [reply src/components/shared/globalTypes.ts:228]: Oh, but instead of throwing the mysterious Error, could just move the return fsObject there and reorder the second if statement to have `!fsObject?.isDirectory` first for short-circuiting. Then, it can only reach the bottom for errors.

Comment (jason2020) [reply src/components/shared/globalTypes.ts:228]: For the case of !fsObject but no onFileFound, this would return the null. But I guess it's okay because that means we wouldn't need to handle noFIleFound for some reason. Actually, now that I think about it, this might be a non-optional param.

File: src/styles/Task.scss (hunk 20 above)
Comment (jason2020) [reply src/styles/Task.scss:5]: Lol, yeah my CSS is a bit sus. Uh, maybe I'll add a TODO to the backlog of changing this `132px` margin for responsive design.

File: src/components/shared/Terminal.tsx (hunk 21 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:276]: Yea, this will be in the backlog tasks PR to handle space at the top (outside of the switch statement). And agreed, arguments should just be parsed into arrays (with spaces as a delimiter between diff args).

File: src/components/shared/Terminal.tsx (hunk 22 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:326]: Agreed, will add 'CLEAR' to a constants file (backlog task)

File: src/components/shared/globalTypes.ts (hunk 18 above)
Comment (jason2020) [reply src/components/shared/globalTypes.ts:215]: hahaha, thank you for catching this! But this raises a good pt about why unit tests should probably be made. It would've caught this error.

File: src/components/shared/globalTypes.ts (hunk 17 above)
Comment (jason2020) [reply src/components/shared/globalTypes.ts:197]: Agreed, index is not used in this loop I think.

File: src/components/shared/globalTypes.ts (hunk 16 above)
Comment (jason2020) [reply src/components/shared/globalTypes.ts:190]: Yup, this simplifies the logic a lot. Makes it way more readable.

File: src/components/shared/globalTypes.ts (hunk 15 above)
Comment (jason2020) [reply src/components/shared/globalTypes.ts:185]: Yup

File: src/components/shared/globalTypes.ts (hunk 13 above)
Comment (jason2020) [reply src/components/shared/globalTypes.ts:156]: lol, yeah this is an issue I'll be addressing soon. should be null.

File: src/components/shared/Terminal.tsx (hunk 23 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:336]: Yeah, will be changing handleChange, handleKeyDown, handleSubmit to match the named functions of the terminal funcs. This should improve code style and make everything more cohesive, good catch.

File: src/components/shared/Terminal.tsx (hunk 4 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:116]: Yeah, I can see how that might be unclear now. Will be adding some commends to better explain on a high-level what these functions do.

File: src/components/shared/Terminal.tsx (hunk 2 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:70]: Yea, this is clearer. A future backlog task would be to refactor and change all instances of fileSystem to be simply root (including changing the name in state).

File: src/components/shared/Terminal.tsx (hunk 1 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:21]: Yup, better practice would be to add these into a constants file. Also, makes it easier to check which operations are supported at a glance. Will do this in a backlog PR.

File: src/components/shared/Terminal.tsx (hunk 6 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:146]: Lol, yeah this part is kinda disturbing. I'll make a change to make it more clear what is happening here.

File: src/components/shared/Terminal.tsx (hunk 5 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:164]: Ohh, that's a good pt. Will do this change, makes for better maintainability of the code base.

File: src/components/shared/Terminal.tsx (hunk 10 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:221]: Yeah, that'll make the code a lot more readable, good suggestion

File: src/components/shared/Terminal.tsx (hunk 11 above)
Comment (jason2020) [reply src/components/shared/Terminal.tsx:230]: Yeah, the performance drawback of indexOf is really small, so I might add this in the future to improve code readability.

File: src/components/shared/globalTypes.ts (hunk 14 above)
Comment (jason2020) [reply src/components/shared/globalTypes.ts:171]: Hmm, I think no because the return is different for these two cases. It's quite similar, but the first is that no such directory exists, but the second is that the newCwd is simply not a director (it's a file with the same path). They use to be treated as the same case, but since the error is slightly different, the if statements should be separate.

**Comment by jason2020** [comment]
@RogueArt Thanks for the helpful suggestions and feedback. As mentioned above, I'll be refactoring the PR into 1 PR for Task refactor and another for the Terminal functionality. Then, I'll make PR(s) for the backlog tasks resolving the issues you mentioned (creating constants, using null instead of undefined, other minor changes, better argument handling). Also most importantly, unit tests are urgently needed. I think maybe I should create a backlog list somewhere.
